File: Python/0015_3Sum/threeSum.py
class Solution:
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        result = []
        for i in range(len(nums)):            
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            left = i + 1
            right = len(nums) - 1            
            
            while left < right:                
                summation = nums[left] + nums[right] + nums[i]
                
                if summation == 0:
                    result.append([nums[i], nums[left], nums[right]])
                    right -= 1
                    left += 1
                    while left < right and nums[left] == nums[left - 1]:
                        left += 1
                    while left < right and nums[right] == nums[right + 1]:
                        right -= 1
                elif summation > 0:
                    right -= 1
                else:
                    left += 1
        return result

File: Csharp/Leetcode/0015_ThreeSum/ThreeSum.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Leetcode.Solutions {
    public IList<IList<int>> ThreeSum(int[] nums) {
        IList<IList<int>> result = new List<IList<int>>();
        Array.Sort(nums);
        
        for (var i = 0; i < nums.Length; i++) {
            if (i != 0 && nums[i] == nums[i - 1]) {
                continue;
            }
            int left = i + 1, right = nums.Length - 1;
            while (left < right) {
                var sum = nums[left] + nums[right];
                if (sum == -nums[i]) {
                    result.Add(new int[]{nums[i], nums[left], nums[right]});
                    do {
                        left++;
                    } while (left < right && nums[left] == nums[left - 1]);                    
                    do {
                        right--;   
                    } while (left < right && nums[right] == nums[right + 1]);
                } else if (sum > -nums[i]) {
                    right--;
                } else {
                    left++;
                }
            }
        }        
        return result;
    }
}
